File: core_libraries/subtrees/preprocessing_research_tool/python/tools/pre_whiten.py
from sklearn.linear_model import LinearRegression
import numpy as np

def pre_whiten(data: np.ndarray):
    """_summary_

    Args:
        data (_type_): _description_
    """
    for i in range(data.shape[1]):
        vals = data[:,i].reshape(-1, 1)
        if np.sum(~np.isnan(vals)) == 0:
            continue
        model = LinearRegression().fit(vals[:-1,:], vals[1:,:])
        E = model.predict(vals[:-1,:]) -vals[1:,:]
        if len(E) < len(vals):
            E = np.concatenate([E,np.nan*np.zeros([len(vals)-len(E),1])])
        data[:,i] = E.reshape(-1)
    
    return data

File: core_libraries/subtrees/preprocessing_research_tool/matlab/tools/other_eeg_processing/pre_whiten.m
function data = pre_whiten(data)


for j = 1:size(data,2)
    vals = data(:,j);    
    
    if sum(~isnan(vals)) == 0
        continue
    end
    
    %{
    mdl = regARIMA('ARLags',1);
    mdl = estimate(mdl,vals,'Display','Off');
    E = infer(mdl,vals);
    %}

    %
    mdl = fitlm(vals(1:end-1),vals(2:end));
    E = mdl.predict(vals(1:end-1)) - vals(2:end);
    %}

    if length(E) < length(vals)
        E = [E;nan(length(vals)-length(E),1)];
    end
    data(:,j) = E;
end



end